packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,239 @@
+
+    @state()
+    private _weakPasswords: ListItem[] = [];
+
+    @state()
+    private _compromisedPasswords: ListItem[] = [];
+
+    @queryAll("pl-audit-list-item")
+    private _countElements: HTMLDivElement[];
+
+    handleRoute() {
+        this.audit();
+    }
+
+    shouldUpdate() {
+        return !!app.account;
+    }
+
+    static styles = [
+        ...View.styles,
+        css`
+            .counts {
+                display: grid;
+                grid-gap: 1em;
+                margin: 1em;
+                grid-template-columns: repeat(auto-fit, minmax(20em, 1fr));
+            }
+
+            @media (max-width: 700px) {
+                .counts {
+                    grid-template-columns: repeat(auto-fit, minmax(15em, 1fr));
+                }
+            }
+        `,
+    ];
+
+    render() {
+        const { _reusedPasswords, _weakPasswords, _compromisedPasswords } = this;
+
+        return html`
+            <div class="fullbleed vertical layout">
+                <header class="padded spacing center-aligning horizontal layout">
+                    <pl-button
+                        class="transparent skinny menu-button header-title"
+                        @click=${() =>
+                            this.dispatchEvent(new CustomEvent("toggle-menu", { composed: true, bubbles: true }))}
+                    >
+                        <div class="half-margined horizontal spacing center-aligning layout text-left-aligning">
+                            <pl-icon icon="shield-check"></pl-icon>
+                            <div class="stretch ellipsis">${$l("Password Audit")}</div>
+                        </div>
+                    </pl-button>
+                </header>
+                <div class="layout padded">
+                    <div class="vertical spacing">
+                        <pl-scroller class="stretch">
+                            <div class="counts">
+                                ${this._renderSection(_reusedPasswords, $l("Reused Passwords"), {
+                                    auditReused: "true",
+                                })}
+                                ${this._renderSection(_weakPasswords, $l("Weak Passwords"), { auditWeak: "true" })}
+                                ${this._renderSection(_compromisedPasswords, $l("Compromised Passwords"), {
+                                    auditCompromised: "true",
+                                })}
+                            </div>
+                        </pl-scroller>
+                    </div>
+                </div>
+            </div>
+        `;
+    }
+
+    async audit() {
+        const reusedPasswords: ListItem[] = [];
+        const weakPasswords: ListItem[] = [];
+        const compromisedPasswords: ListItem[] = [];
+
+        const { vaults } = this.state;
+
+        const firstMatchPerPasswordHash: {
+            [passwordHash: string]: { item: VaultItem; vault: Vault; passwordField: Field };
+        } = {};
```

Comment:
Why not use a `Map<string, {...}>` for this?

Reply:
I've never found it to be much more advantageous than a plain object for <1k entries, but I see no harm here, and thinking about it now, I suppose there will be many people with thousands of items. Good suggestion, thanks!

Reply:
I don't think there is a measurable difference in performance, but I think it's semantically more appropriate. Being able to use plain objects as dictionaries is a neat feature in javascript but for me it always felt kinda wrong 😅. Using `Map` makes it clear what kind of data structure we're creating and for what purpose. This is, once again, just a style question though, so I don't feel too strongly either way.